### PR TITLE
DEVPROD-7529 Allow patch trigger aliases to specify revision

### PR DESCRIPTION
**Pull request by hadjri**
DEVPROD-7529

### Description
This adds a new field on the patch trigger definition that specifies which revision to base the downstream patch off of, rather than having it always default to the latest revision for the downstream patch.

The team requesting / using this will only be setting this via API, should we keep this feature API-only?

### Testing
Confirmed expected behavior in staging + unit test

### CODE REVIEW

File: rest/model/project.go [hunk 1]
```diff
@@ -106,6 +106,8 @@ type APIPatchTriggerDefinition struct {
 	// Name of the module corresponding to the upstream project in the
 	// downstream project's YAML.
 	ParentAsModule *string `json:"parent_as_module,omitempty"`
+	// The revision at which to create the downstream patch.
+	DownstreamRevision *string `json:"downstream_revision,omitempty"`
```
Comment (Kimchelly) [inline rest/model/project.go:110]: It looks like this API model is also used by GQL resolvers - just to check, if a user does something like save the triggers in project settings, would it clear this field or preserve it? I just fixed an issue where GQL didn't propagate a field and it accidentally cleared the setting each time it saved the project settings section, so just wanted to double-check.

File: units/patch_intent_test.go [hunk 2]
```diff
@@ -1626,6 +1626,74 @@ tasks:
...
+	}
+	s.NoError(p.Insert())
+
+	u := &user.DBUser{
+		Id: evergreen.ParentPatchUser,
+	}
+	s.NoError(u.Insert())
+
+	projectRef, err := model.FindBranchProjectRef(s.project)
+	s.NotNil(projectRef)
```
Comment (Kimchelly) [inline units/patch_intent_test.go:1674]: `s.Require().NotNil(projectRef)` to avoid nil panics.

File: units/patch_intent_test.go [hunk 3]
```diff
@@ -1626,6 +1626,74 @@ tasks:
...
+	s.NoError(err)
+	s.Require().Len(projectRef.PatchTriggerAliases, 1)
+	projectRef.PatchTriggerAliases[0].DownstreamRevision = "abc"
+
+	s.Len(p.Triggers.ChildPatches, 0)
```
Comment (Kimchelly) [inline units/patch_intent_test.go:1679]: Is this check relevant to the test? It seems like `p` hasn't been used so far, so I would expect that `p` would still have no child patches just from the assignment on L1660-1664.

File: units/patch_intent_test.go [hunk 4]
```diff
@@ -1626,6 +1626,74 @@ tasks:
...
+	s.NoError(ProcessTriggerAliases(s.ctx, p, projectRef, s.env, []string{"patch-alias"}))
+
+	dbPatch, err := patch.FindOneId(p.Id.Hex())
+	s.NoError(err)
+	s.Require().NotZero(dbPatch)
+	s.Equal(p.Triggers.ChildPatches, dbPatch.Triggers.ChildPatches)
+
+	s.Require().NotEmpty(dbPatch.Triggers.ChildPatches)
```
Comment (Kimchelly) [inline units/patch_intent_test.go:1687]: Is this only expecting non-empty or is it specifically expecting exactly 1 child patch? Might be clearer to do `s.Require().Len` if so.

**Review by Kimchelly — CHANGES_REQUESTED** [review]
I'm a bit confused about the intended workflow here because the downstream revision is set on the project ref's trigger definition - how/when do users set the downstream revision on the project ref for each trigger? Is the expected workflow to programmatically set the latest revision on the project ref's trigger definition within the task, and then keep bumping it each time they run another trigger? A quick explanation or example workflow of how it'll be used would be helpful!

Re the API thing, since this is the only time we've heard this request and it seems quite specific, I think just supporting the user's preferred way of setting the downstream revision is good enough. We can always add support for other methods to set it in the future if someone asks.

File: units/patch_intent.go [hunk 5]
```diff
@@ -1170,6 +1173,26 @@ func (j *patchIntentProcessor) buildTriggerPatchDoc(patchDoc *patch.Patch) (*mod
...
+		v, err = model.VersionFindOne(model.BaseVersionByProjectIdAndRevision(patchDoc.Project, patchDoc.Triggers.DownstreamRevision))
+		if err != nil {
+			return nil, nil, nil, errors.Wrapf(err, "getting version at revision '%s'", patchDoc.Triggers.DownstreamRevision)
+		}
+		project, pp, err = model.FindAndTranslateProjectForVersion(ctx, evergreen.GetEnvironment().Settings(), v, true)
+	} else {
+		v, project, pp, err = model.FindLatestVersionWithValidProject(patchDoc.Project, true)
+	}
+	if err != nil {
+		return nil, nil, nil, errors.Wrapf(err, "getting downstream version to use for patch '%s'", patchDoc.Id.Hex())
```
[the remark this answers is not in the document]
Comment (hadjri) [reply units/patch_intent.go:1191]: Sure, updated

File: units/patch_intent_test.go (hunk 4 above)
Comment (hadjri) [reply units/patch_intent_test.go:1687]: Explicitly 1, updated

File: units/patch_intent_test.go (hunk 3 above)
Comment (hadjri) [reply units/patch_intent_test.go:1679]: It can be removed

**Comment by hadjri** [comment]
> I'm a bit confused about the intended workflow here because the downstream revision is set on the project ref's trigger definition - how/when do users set the downstream revision on the project ref for each trigger? Is the expected workflow to programmatically set the latest revision on the project ref's trigger definition within the task, and then keep bumping it each time they run another trigger? A quick explanation or example workflow of how it'll be used would be helpful!

Their workflow is that they will be programmatically updating a server revision to use for downstream patches on the mongot project every time some cron job runs.

> Re the API thing, since this is the only time we've heard this request and it seems quite specific, I think just supporting the user's preferred way of setting the downstream revision is good enough. We can always add support for other methods to set it in the future if someone asks.

SGTM!

File: rest/model/project.go (hunk 1 above)
Comment (hadjri) [reply rest/model/project.go:110]: That's a great catch -- thank you for flagging. I created a [Spruce PR](https://github.com/evergreen-ci/evergreen/pull/8012) as that's the only way to really get around this.

**Review by Kimchelly — COMMENTED** [review]
~I think you forgot to push your changes?~ JK, they just pushed, never mind!

File: units/patch_intent.go [hunk 6]
```diff
@@ -1183,10 +1183,16 @@ func fetchTriggerProjectInfo(ctx context.Context, patchDoc *patch.Patch) (*model
...
+		if v == nil {
+			return nil, nil, nil, errors.Errorf("version at revision '%s' not found", patchDoc.Triggers.DownstreamRevision)
+		}
 		project, pp, err = model.FindAndTranslateProjectForVersion(ctx, evergreen.GetEnvironment().Settings(), v, true)
-	} else {
-		v, project, pp, err = model.FindLatestVersionWithValidProject(patchDoc.Project, true)
+		if err != nil {
+			return nil, nil, nil, errors.Wrapf(err, "getting downstream version at revision '%s' to use for patch '%s'", patchDoc.Triggers.DownstreamRevision, patchDoc.Id.Hex())
+		}
+		return v, project, pp, nil
```
Comment (Kimchelly) [inline units/patch_intent.go:1193]: Since the `if` block returns at the end, the `var` declarations at the top aren't needed anymore, and the `else` can be removed since everything past L1193 is the same as an `else` block anyways.

**Review by Kimchelly — APPROVED** [review]
LGTM mod one simplification!
